addons/udes_stock/models/stock_move.py

```diff
@@ -310,13 +321,64 @@ def _create_moves_for_push(self, push, move_lines):
             created_moves |= move.copy(move_vals)
         return created_moves
 
+    def group_by_key(self):
+        # force recompute on u_grouping_key so we have an up-to-date key:
+        self._fields['u_grouping_key'].compute_value(self)
+
+        if any(pt.u_move_key_format is False
+               for pt in self.mapped('picking_id.picking_type_id')):
+            raise UserError(_("Cannot group move lines when their picking type"
+                              "has no grouping key set."))
+
+        return self.groupby(lambda ml: ml.u_grouping_key)
+
+    def refactor_action_group_by_move_key(self):
+        """
+        group the moves by the splitting criteria
+        for each resulting group of stock.moves:
+            create a new picking
+            split any stock.move records that are only partially covered by the
+                group of stock.moves
+            attach the stock.moves to the new picking.
+        """
+        Picking = self.env['stock.picking']
+
+        picking_type = self.mapped('picking_type_id')
+        picking_type.ensure_one()
+
+        if not picking_type.u_move_key_format:
+            return
+
+        pickings = self.mapped('picking_id')
+
+        for key, move_group in self.group_by_key():
+
+            if len(move_group.mapped('location_id')) > 1 or \
+                    len(move_group.mapped('location_dest_id')) > 1:
+                raise UserError(
+                    _('Move grouping has generated a group of'
+                      'moves that has more than one source or '
+                      'destination location. Aborting. key: "%s", '
+                      'location_ids: "%s", location_dest_ids: "%s"'
+                      '') % (key, move_group.mapped('location_id'),
+                             move_group.mapped('location_dest_id')))
+
+            Picking._new_picking_for_group(key, move_group)
+
+        empty_picks = pickings.filtered(lambda p: len(p.move_lines) == 0)
+        if empty_picks:
+            _logger.info(_("Cancelling empty picks after splitting."))
+            # action_cancel does not cancel a picking with no moves.
+            empty_picks.write({
+                'state': 'cancel',
+                'is_locked': True
+            })
+
```

Comment:
Need to return the union of all `move_group`s at the end

Reply:
wouldn't that just be self?